service/src/event/ordering_task.rs

```diff
@@ -365,112 +434,96 @@ impl OrderingState {
         &mut self,
         pool: &SqlitePool,
         max_iterations: usize,
-    ) -> Result<()> {
-        let mut cnt = 0;
-        let mut offset: usize = 0;
-        while cnt < max_iterations {
-            cnt += 1;
-            let (new, found) = self
-                .add_undelivered_batch(pool, offset, DELIVERABLE_EVENTS_BATCH_SIZE)
-                .await?;
-            if new == 0 {
+        batch_size: u32,
+    ) -> Result<usize> {
+        let mut iter_cnt = 0;
+        let mut event_cnt = 0;
+        let mut highwater = 0;
+        while iter_cnt < max_iterations {
+            iter_cnt += 1;
+            let (undelivered, new_hw) =
+                CeramicOneEvent::undelivered_with_values(pool, batch_size.into(), highwater)
+                    .await?;
+            highwater = new_hw;
+            if undelivered.is_empty() {
                 break;
             } else {
                 // We can start processing and we'll follow the stream history if we have it. In that case, we either arrive
                 // at the beginning and mark them all delivered, or we find a gap and stop processing and leave them in memory.
                 // In this case, we won't discover them until we start running recon with a peer, so maybe we should drop them
                 // or otherwise mark them ignored somehow.
-                self.process_events(pool, None).await?;
-                if new < DELIVERABLE_EVENTS_BATCH_SIZE {
+                let found_all = undelivered.len() < batch_size as usize;
```

Comment:
why isn't this `undelivered.len() == 0`?  What happens if you have a handful of leftover undelivered events but not enough to fill a batch?

Reply:
a handful of events that don't fill a batch is this `found_all` (now `found_everything`) and we will exit the loop after processing them. I modified it to only have one break at the end rather than two under conditionals.